### PR TITLE
Adds documentation around the halibut protocol.

### PR DESCRIPTION
**Pull request by LukeButters**
# Background

Illustrate the halibut protocol data exchanged.

[SC-38515](https://app.shortcut.com/octopusdeploy/story/38515/document-halibut-protocol)

### CODE REVIEW

File: docs/halibut-protocol.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+## Polling client protocol data exchange
+
+![Polling client protocol data exchange](images/pollingprotocoldata.png)
+
+## Request and Response message format
+
+Request and Response messages are sent and received in the same format with only the data within the BSON section changing to denote the type of message.
+
+![Request/Response message format](images/message-format.png)
```
Comment (geofflamrock) [inline docs/halibut-protocol.md:24]: These images don't appear to be rendering all that well when I look at the [branch directly in GitHub](https://github.com/OctopusDeploy/Halibut/blob/luke/illustrate-the-halibut-protocol/docs/halibut-protocol.md), I wonder if they could do with a white background?

File: docs/halibut-protocol.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Halibut Protocol data exchange
+
+When invoking a remote method in both polling and listening mode, the same general process is followed:
+
+ - The Server sends a `Request` message containing the method to execute and data.
```
[the remark this answers is not in the document]
Comment (LukeButters) [reply docs/halibut-protocol.md:5]: I think defining them upfront makes sense.

File: docs/halibut-protocol.md (hunk 1 above)
Comment (LukeButters) [reply docs/halibut-protocol.md:24]: I see them with a white background...

File: docs/halibut-protocol.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# Halibut Protocol data exchange
```
[the remark this answers is not in the document]
Comment (LukeButters) [reply docs/halibut-protocol.md:1]: Updated to make mention of the control messages during initiation as well as some hints at what the messages are trying to signal.

File: docs/halibut-protocol.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+![Listening client protocol data exchange](images/listeningprotocoldata.png)
+
+## Polling client protocol data exchange
+
+![Polling client protocol data exchange](images/pollingprotocoldata.png)
+
+## Request and Response message format
+
+Request and Response messages are sent and received in the same format with only the data within the BSON section changing to denote the type of message.
+
```
[the remark this answers is not in the document]
Comment (LukeButters) [reply docs/halibut-protocol.md:34]: How is that?

File: docs/halibut-protocol.md (hunk 1 above)
Comment (LukeButters) [reply docs/halibut-protocol.md:35]: Although for me they are blurry:

image

File: docs/halibut-protocol.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
+# Halibut Protocol data exchange
+
+This document outlines the data that is exchanged between the **Client** and **Service**. In this document:
+
+ - **Client** is what makes the remote procedure calls e.g. Octopus.
+ - **Service** is what execiutes the methods e.g. Tentacle.
```
Comment (LukeButters) [inline docs/halibut-protocol.md:6]: These terms are lifted from the Halibut readme.

File: docs/halibut-protocol.md (hunk 1 above)
Comment (geofflamrock) [reply docs/halibut-protocol.md:35]: I think they have a transparent background, I'm using a dark theme in GitHub so the background appears dark and some of the words are hard to read.

image

Comment (LukeButters) [reply docs/halibut-protocol.md:35]: oh boy

File: docs/halibut-protocol.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+## Polling Service protocol data exchange
+
+When the **Service** is in  **Polling** mode, the **Service** identifies itself as `MX-SUBSCRIBER 1.0` while the `Client` identifies itself as `MX-SERVER 1.0`.
+
+![Polling client protocol data exchange](images/pollingprotocoldata.png)
+
+## Request and Response message format
+
+The message format is always a Zipped BSON representation of either the [Request](../source/Halibut/Transport/Protocol/RequestMessage.cs) or [Response](../source/Halibut/Transport/Protocol/ResponseMessage.cs) message, followed by multiple [DataStream](../source/Halibut/DataStream.cs). DataStreams is the way Halibut sends large amounts of data. DataStreams are not compressed and are send one by one after the compressed BSON. DataStreams are referenced by GUID in the BSON section and labelled with a GUID in the DataStream section.
```
Comment (geofflamrock) [inline docs/halibut-protocol.md:33]:
```suggestion
The message format is always a Zipped BSON representation of either the [Request](../source/Halibut/Transport/Protocol/RequestMessage.cs) or [Response](../source/Halibut/Transport/Protocol/ResponseMessage.cs) message, followed by zero or more [DataStream]s(../source/Halibut/DataStream.cs). A `DataStream` represents data that should not be serialized as part of a message, for example a file to be transferred. They can be sent in either a request or a response. `DataStream`s are transferred as raw bytes in the TCP stream (i.e. they are not compressed) and are sent sequentially after the compressed BSON of the request/response. Each `DataStream` has a unique GUID which is referenced in the request/response so that it can be used by calling code.
```

File: docs/halibut-protocol.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (geofflamrock) [reply docs/halibut-protocol.md:34]: Mostly looks good, I've made a couple of suggestions inline for you to consider.

File: docs/halibut-protocol.md (hunk 1 above)
Comment (geofflamrock) [reply docs/halibut-protocol.md:35]: If you re-export the pictures with a white background hopefully it should appear ok for both light and dark themes?

File: docs/halibut-protocol.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+This document outlines the data that is exchanged between the **Client** and **Service**. In this document:
+
+ - **Client** is what makes the remote procedure calls e.g. Octopus Server.
+ - **Service** is what executes the methods e.g. Tentacle. 
+
+
+When invoking a remote method in both polling and listening mode, the same general process is followed:
+
+1.  After the TCP connection is intiated, both server and client identify themselves to each other through a set of control messages.
```
Comment (geofflamrock) [inline docs/halibut-protocol.md:11]: It might be worth mentioning that the identification is different for listening vs polling tentacles and adding a link to those sections.

File: docs/halibut-protocol.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+2. The Client sends a `Request` message containing the method to execute and data.
+3. The Service executes the method and sends the result in a `Response` message.
+4. The Service sends a `NEXT` control message, signaling it is ready for the next action.
+5. The Client sends a `Proceed` control message, signaling to the Service to be ready for another `Request`.
```
Comment (geofflamrock) [inline docs/halibut-protocol.md:15]: The casing is different here compared to `NEXT` and `END`, is that how Halibut works?

File: docs/halibut-protocol.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (geofflamrock) [reply docs/halibut-protocol.md:1]: Starting to look good, although I'm finding the numbered steps with the "Steps 2 to 5 are repeated..." slightly hard to follow. 

I wonder if it might be helpful to split the numbered list into 3 sub-sections to help show the various phases of the connection:
- Identification
- Message exchange
- Ending a connection

Having separate sub-sections might allow more room to expand on some things like the difference between graceful and non-graceful termination of the connection (e.g. do both parties send the `END` message?)

What do you think?

**Review by geofflamrock — COMMENTED** [review]
This is starting to look good 👍 I've left a couple of more thoughts on potential improvements

File: docs/halibut-protocol.md (hunk 8 above)
Comment (LukeButters) [reply docs/halibut-protocol.md:15]: nice catch

File: docs/halibut-protocol.md (hunk 1 above)
Comment (LukeButters) [reply docs/halibut-protocol.md:35]: try now?

File: docs/halibut-protocol.md (hunk 7 above)
Comment (LukeButters) [reply docs/halibut-protocol.md]: Done

File: docs/halibut-protocol.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+ - **Client** is what makes the remote procedure calls e.g. Octopus Server.
+ - **Service** is what executes the methods e.g. Tentacle. 
+
+
+When invoking a remote method in both polling and listening mode, the `Client` and `Service` go through three stages.
+
+ ### Stage 1 - Identification
+ 
+  A TCP connection is established and both server and client identify themselves, though a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.
```
Comment (veochen-octopus) [inline docs/halibut-protocol.md:13]:
```suggestion
  A TCP connection is established and both server and client identify themselves, through a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.
```

File: docs/halibut-protocol.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+The steps are repeated for both Listening and Polling mode until either Client or Service begins the Ending a connection stage.
+
+### Stage 3 - Ending a connection
+
+The final stage can be entered by either the Client or Service terminating the TCP connection or sending a `END` control message. Once this stage is reached no more requests will be processed in the current TCP connection.
+
+## Listening Service protocol data exchange
+
+When the **Service** is in  **Listening** mode, the **Client** it identifies itself as `MX-CLIENT 1.0` while the `Service` identifies itself as `MX-SERVER 1.0`.
```
Comment (veochen-octopus) [inline docs/halibut-protocol.md:32]:
```suggestion
When the **Service** is in  **Listening** mode, the **Client** identifies itself as `MX-CLIENT 1.0` while the `Service` identifies itself as `MX-SERVER 1.0`.
```

File: docs/halibut-protocol.md [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+### Stage 2 - Message exchange
+
+Now that the connections are established, the Service will wait for requests from the Client. In both polling and listening the following steps are repeated in order:
+
+1. The Client sends a `Request` message containing the method to execute and data.
+2. The Service executes the method and sends the result in a `Response` message.
+3. The Service sends a `NEXT` control message, signaling it is ready for the next action.
+4. The Client sends a `PROCEED` control message, signaling to the Service to be ready for another `Request`.
+
+The steps are repeated for both Listening and Polling mode until either Client or Service begins the Ending a connection stage.
```
Comment (veochen-octopus) [inline docs/halibut-protocol.md:24]: Is it worth mentioning that during polling, the `NEXT`/`PROCEED` cycle repeats itself periodically when no requests are being sent?

**Review by veochen-octopus — COMMENTED** [review]
Thanks @LukeButters. This will be incredibly useful. Just a few minor things from me.

File: docs/halibut-protocol.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (LukeButters) [reply docs/halibut-protocol.md:1]: I tried to do this, did it work?

File: docs/halibut-protocol.md (hunk 11 above)
Comment (LukeButters) [reply docs/halibut-protocol.md:24]: Done under the polling section

File: docs/halibut-protocol.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (geofflamrock) [reply docs/halibut-protocol.md:1]: Yep that reads much better, thanks.

File: docs/halibut-protocol.md (hunk 1 above)
Comment (geofflamrock) [reply docs/halibut-protocol.md:48]: Yep looking good now, thanks

**Review by geofflamrock — COMMENTED** [review]
Looks good to me👍I'll wait for others to take a look before approving.

**Review by veochen-octopus — APPROVED** [review]
This is looking really good. Thanks for putting this together.

File: docs/halibut-protocol.md [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+ - **Client** is what makes the remote procedure calls e.g. Octopus Server.
+ - **Service** is what executes the methods e.g. Tentacle. 
+
+
+When invoking a remote method in both polling and listening mode, the `Client` and `Service` go through three stages.
+
+ ### Stage 1 - Identification
+ 
+  A TCP connection is established and both server and client identify themselves, through a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.
```
Comment (nathanwoctopusdeploy) [inline docs/halibut-protocol.md:13]: server -> service

Comment (LukeButters) [reply docs/halibut-protocol.md]: fixed

File: docs/halibut-protocol.md [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+When the **Service** is in  **Polling** mode, the **Service** identifies itself as `MX-SUBSCRIBER 1.0` while the `Client` identifies itself as `MX-SERVER 1.0`.
+
+Under polling mode, unlike listening, the Client will periodically send a no-op NUL Request to the Service to keep the TCP connection alive.
+
+![Polling client protocol data exchange](images/pollingprotocoldata.png)
+
+## Request and Response message format
+
+The message format is always a Zipped BSON representation of either the [Request](../source/Halibut/Transport/Protocol/RequestMessage.cs) or [Response](../source/Halibut/Transport/Protocol/ResponseMessage.cs) message, followed by zero or more [DataStream]s(../source/Halibut/DataStream.cs). A `DataStream` represents data that should not be serialized as part of a message, for example a file to be transferred. They can be sent in either a request or a response. `DataStream`s are transferred as raw bytes in the TCP stream (i.e. they are not compressed) and are sent sequentially after the compressed BSON of the request/response. Each `DataStream` has a unique GUID which is referenced in the request/response so that it can be used by calling code.
```
Comment (nathanwoctopusdeploy) [inline docs/halibut-protocol.md:46]: [DataStream]s(../source/Halibut/DataStream.cs) is not linking tot he file due to the rouge s

Comment (LukeButters) [reply docs/halibut-protocol.md]: Fixed

**Review by nathanwoctopusdeploy — APPROVED** [review]
Awesome job, super clear 👍 

One suggestion would be to pull the whimsical drawings out into there own whimsical diagram (if they aren't already) and link to it so future updates can be made if needed / the source drawings can be used by people if they wanted.

**Comment by LukeButters** [comment]
@nathanwoctopusdeploy done

File: docs/halibut-protocol.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
+# Halibut Protocol data exchange
```
Comment (pawelpabich) [inline docs/halibut-protocol.md:1]: Consider mentioning that the protocol is inspired by JSON-RPC and adding a link to https://www.codeproject.com/Articles/554692/Halibut-a-secure-communication-stack-for-NET-Mono to provide some historical context.

File: docs/halibut-protocol.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+ ### Stage 1 - Identification
+ 
+  A TCP connection is established and both Service and client identify themselves, through a series of control messages. The control messages sent vary depending on if [listening](#listening-service-protocol-data-exchange) or [polling](#polling-service-protocol-data-exchange) mode is used.
+
+### Stage 2 - Message exchange
+
+Now that the connections are established, the Service will wait for requests from the Client. In both polling and listening the following steps are repeated in order:
+
+1. The Client sends a `Request` message containing the method to execute and data.
```
Comment (pawelpabich) [inline docs/halibut-protocol.md:19]: `data` is a bit overload term in this context. What about `message containing the method to execute, its parameters and 0 or more data streams`

File: docs/halibut-protocol.md [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+2. The Service executes the method and sends the result in a `Response` message.
```
Comment (pawelpabich) [inline docs/halibut-protocol.md:20]: Similar to above, "The Service executes the method and sends a `Response` message containing the result and 0 or more data streams`

**Review by pawelpabich — COMMENTED** [review]
This will make a lot of confused people happy :).

File: docs/halibut-protocol.md (hunk 14 above)
Comment (LukeButters) [reply docs/halibut-protocol.md:1]: added
